File: src/系统/00．核心系统/01．事件中心/07A．单位排泄.lua
local ____lualib = require("lualib_bundle")
local __TS__Delete = ____lualib.__TS__Delete
local __TS__ArraySplice = ____lualib.__TS__ArraySplice
local ____exports = {}
---
-- @noSelfInFile
local jass = require("jass.common")
local ____require_result_0 = require("系统.00．核心系统.01．事件中心.07．单位死亡事件中心")
local registerDeathListener = ____require_result_0.registerDeathListener
local ____require_result_1 = require("系统.00．核心系统.05．中心计时器")
local onTick10ms = ____require_result_1.onTick10ms
local getServerTime = ____require_result_1.getServerTime
local GetHandleId = jass.GetHandleId
local GetUnitState = jass.GetUnitState
local IsUnitType = jass.IsUnitType
local RemoveUnit = jass.RemoveUnit
local _____5355_4F4D_6392_6CC4_5EF6_8FDF_6BEB_79D2 = 20000
local _____6B7B_4EA1_751F_547D_9608_503C = 0.405
local _____5DF2_767B_8BB0_5355_4F4D_8868 = {}
local _____5DF2_8FDB_5165_6392_6CC4_961F_5217 = {}
local _____5F85_6392_6CC4_5355_4F4D_5217_8868 = {}
local _____5DF2_6CE8_518C_5355_4F4D_6B7B_4EA1_76D1_542C = false
local _____5DF2_6CE8_518C_5355_4F4D_6392_6CC4Tick = false
local function _____5355_4F4D_662F_5426_4ECD_4E3A_5C38_4F53(unit)
    if unit == nil or unit == 0 then
        return false
    end
    if IsUnitType(unit, jass.UNIT_TYPE_DEAD) then
        return true
    end
    return GetUnitState(unit, jass.UNIT_STATE_LIFE) <= _____6B7B_4EA1_751F_547D_9608_503C
end
local function ____on_5355_4F4D_6392_6CC4Tick()
    local now = getServerTime()
    do
        local i = #_____5F85_6392_6CC4_5355_4F4D_5217_8868 - 1
        while i >= 0 do
            do
                local entry = _____5F85_6392_6CC4_5355_4F4D_5217_8868[i + 1]
                if now < entry.dueTimeMs then
                    goto __continue7
                end
                __TS__Delete(_____5DF2_8FDB_5165_6392_6CC4_961F_5217, entry.handleId)
                if _____5355_4F4D_662F_5426_4ECD_4E3A_5C38_4F53(entry.unit) then
                    RemoveUnit(entry.unit)
                end
                __TS__ArraySplice(_____5F85_6392_6CC4_5355_4F4D_5217_8868, i, 1)
            end
            ::__continue7::
            i = i - 1
        end
    end
end
local function ____on_5DF2_767B_8BB0_5355_4F4D_6B7B_4EA1(dyingUnit)
    if dyingUnit == nil or dyingUnit == 0 then
        return
    end
    local handleId = GetHandleId(dyingUnit)
    if handleId == 0 then
        return
    end
    if _____5DF2_767B_8BB0_5355_4F4D_8868[handleId] ~= true then
        return
    end
    __TS__Delete(_____5DF2_767B_8BB0_5355_4F4D_8868, handleId)
    if _____5DF2_8FDB_5165_6392_6CC4_961F_5217[handleId] == true then
        return
    end
    _____5DF2_8FDB_5165_6392_6CC4_961F_5217[handleId] = true
    _____5F85_6392_6CC4_5355_4F4D_5217_8868[#_____5F85_6392_6CC4_5355_4F4D_5217_8868 + 1] = {
        handleId = handleId,
        unit = dyingUnit,
        dueTimeMs = getServerTime() + _____5355_4F4D_6392_6CC4_5EF6_8FDF_6BEB_79D2
    }
end
local function _____786E_4FDD_5355_4F4D_6392_6CC4_521D_59CB_5316()
    if not _____5DF2_6CE8_518C_5355_4F4D_6B7B_4EA1_76D1_542C then
        _____5DF2_6CE8_518C_5355_4F4D_6B7B_4EA1_76D1_542C = true
        registerDeathListener(____on_5DF2_767B_8BB0_5355_4F4D_6B7B_4EA1)
    end
    if not _____5DF2_6CE8_518C_5355_4F4D_6392_6CC4Tick then
        _____5DF2_6CE8_518C_5355_4F4D_6392_6CC4Tick = true
        onTick10ms(____on_5355_4F4D_6392_6CC4Tick)
    end
end
____exports["登记单位排泄"] = function(unit)
    if unit == nil or unit == 0 then
        return unit
    end
    _____786E_4FDD_5355_4F4D_6392_6CC4_521D_59CB_5316()
    local handleId = GetHandleId(unit)
    if handleId == 0 then
        return unit
    end
    _____5DF2_767B_8BB0_5355_4F4D_8868[handleId] = true
    __TS__Delete(_____5DF2_8FDB_5165_6392_6CC4_961F_5217, handleId)
    return unit
end
____exports["取消单位排泄登记"] = function(unit)
    if unit == nil or unit == 0 then
        return unit
    end
    local handleId = GetHandleId(unit)
    if handleId == 0 then
        return unit
    end
    __TS__Delete(_____5DF2_767B_8BB0_5355_4F4D_8868, handleId)
    __TS__Delete(_____5DF2_8FDB_5165_6392_6CC4_961F_5217, handleId)
    do
        local i = #_____5F85_6392_6CC4_5355_4F4D_5217_8868 - 1
        while i >= 0 do
            if _____5F85_6392_6CC4_5355_4F4D_5217_8868[i + 1].handleId == handleId then
                __TS__ArraySplice(_____5F85_6392_6CC4_5355_4F4D_5217_8868, i, 1)
            end
            i = i - 1
        end
    end
    return unit
end
____exports["立即移除单位并取消排泄登记"] = function(unit)
    if unit == nil or unit == 0 then
        return
    end
    ____exports["取消单位排泄登记"](unit)
    RemoveUnit(unit)
end
____exports["注册单位排泄监听"] = ____exports["登记单位排泄"]
____exports["注销单位排泄监听"] = ____exports["取消单位排泄登记"]
____exports["立即移除单位并注销排泄监听"] = ____exports["立即移除单位并取消排泄登记"]
return ____exports

File: TS/系统/00．核心系统/01．事件中心/07A．单位排泄.ts
/** @noSelfInFile */

const jass = require("jass.common") as any;

const { registerDeathListener } = require("系统.00．核心系统.01．事件中心.07．单位死亡事件中心") as {
  registerDeathListener: (this: void, callback: (this: void, dyingUnit: any, killingUnit: any) => void) => void;
};
const { onTick10ms, getServerTime } = require("系统.00．核心系统.05．中心计时器") as {
  onTick10ms: (this: void, callback: (this: void) => void) => void;
  getServerTime: (this: void) => number;
};

const GetHandleId = jass.GetHandleId as (this: void, handle: any) => number;
const GetUnitState = jass.GetUnitState as (this: void, whichUnit: any, whichUnitState: any) => number;
const IsUnitType = jass.IsUnitType as (this: void, whichUnit: any, whichUnitType: any) => boolean;
const RemoveUnit = jass.RemoveUnit as (this: void, whichUnit: any) => void;

const 单位排泄延迟毫秒 = 20000;
const 死亡生命阈值 = 0.405;

interface 待排泄单位 {
  handleId: number;
  unit: any;
  dueTimeMs: number;
}

const 已登记单位表: Record<number, true | undefined> = {};
const 已进入排泄队列: Record<number, true | undefined> = {};
const 待排泄单位列表: 待排泄单位[] = [];

let 已注册单位死亡监听 = false;
let 已注册单位排泄Tick = false;

function 单位是否仍为尸体(this: void, unit: any): boolean {
  if (unit == null || unit === 0) return false;
  if (IsUnitType(unit, jass.UNIT_TYPE_DEAD)) return true;
  return GetUnitState(unit, jass.UNIT_STATE_LIFE) <= 死亡生命阈值;
}

function on单位排泄Tick(this: void): void {
  const now = getServerTime();
  for (let i = 待排泄单位列表.length - 1; i >= 0; i--) {
    const entry = 待排泄单位列表[i];
    if (now < entry.dueTimeMs) continue;

    delete 已进入排泄队列[entry.handleId];
    if (单位是否仍为尸体(entry.unit)) {
      RemoveUnit(entry.unit);
    }
    待排泄单位列表.splice(i, 1);
  }
}

function on已登记单位死亡(this: void, dyingUnit: any): void {
  if (dyingUnit == null || dyingUnit === 0) return;
  const handleId = GetHandleId(dyingUnit);
  if (handleId === 0) return;
  if (已登记单位表[handleId] !== true) return;

  delete 已登记单位表[handleId];
  if (已进入排泄队列[handleId] === true) return;
  已进入排泄队列[handleId] = true;
  待排泄单位列表.push({
    handleId,
    unit: dyingUnit,
    dueTimeMs: getServerTime() + 单位排泄延迟毫秒,
  });
}

function 确保单位排泄初始化(this: void): void {
  if (!已注册单位死亡监听) {
    已注册单位死亡监听 = true;
    registerDeathListener(on已登记单位死亡);
  }
  if (!已注册单位排泄Tick) {
    已注册单位排泄Tick = true;
    onTick10ms(on单位排泄Tick);
  }
}

export function 登记单位排泄(unit: any): any {
  if (unit == null || unit === 0) return unit;
  确保单位排泄初始化();
  const handleId = GetHandleId(unit);
  if (handleId === 0) return unit;

  // 句柄可能被引擎复用；重新登记时必须先清掉同 handleId 的旧延迟排泄条目，
  // 否则旧尸体的排泄任务可能会把新创建的单位误删掉。
  for (let i = 待排泄单位列表.length - 1; i >= 0; i--) {
    if (待排泄单位列表[i].handleId === handleId) {
      待排泄单位列表.splice(i, 1);
    }
  }

  已登记单位表[handleId] = true;
  delete 已进入排泄队列[handleId];
  return unit;
}

export function 取消单位排泄登记(unit: any): any {
  if (unit == null || unit === 0) return unit;
  const handleId = GetHandleId(unit);
  if (handleId === 0) return unit;

  delete 已登记单位表[handleId];
  delete 已进入排泄队列[handleId];
  for (let i = 待排泄单位列表.length - 1; i >= 0; i--) {
    if (待排泄单位列表[i].handleId === handleId) {
      待排泄单位列表.splice(i, 1);
    }
  }
  return unit;
}

export function 立即移除单位并取消排泄登记(unit: any): void {
  if (unit == null || unit === 0) return;
  取消单位排泄登记(unit);
  RemoveUnit(unit);
}

// 兼容旧命名：这里不是“额外注册事件监听”，只是登记到单位排泄表
export const 注册单位排泄监听 = 登记单位排泄;
export const 注销单位排泄监听 = 取消单位排泄登记;
export const 立即移除单位并注销排泄监听 = 立即移除单位并取消排泄登记;
